### DISCUSSION TITLE
Gunicorn / Flask / MySQL - Concurrency question & issues - Data only available on some workers

### DISCUSSION THREAD

**Discussion by zlshames** [sqlalchemy/sqlalchemy#8794] [Usage Questions] [answered]
Hi All,

I currently have a project that utilizes Flask, SQLAlchemy, MySQL, and Gunicorn as the backend stack. My issue arises the more workers I enable for Gunicorn. Basically, data written by 1 worker may not be available in a SQLAlchemy query within another worker. Despite `commit` being used, and the data is viewable in the database. I'll give you a scenario:

1. Gunicorn is run with 4 workers (A, B, C, and D)
2. I use an API call to "login" a user, which creates a token in the database. This hits worker A.
3. I attempt to access an endpoint using the token that was generated in worker A. This hits worker B.
    - I get an Unauthorized error because worker B's query returns no data.
4. I attempt to access an endpoint using the token that was generated in worker A. This hits worker C.
    - I _do not_ get an Unauthorized error because worker C's query somehow is able to see the data written by worker A.
5. I attempt to access an endpoint using the token that was generated in worker A. This hits worker D.
    - I get an Unauthorized error because worker D's query returns no data.

As you can see, data written to the database via `commit` are only queryable on some of the workers (processes), and I don't understand why that would be. Here is a bit of sample code for how I instantiate my DB engine:

```
self.engine = create_engine("mysql+pymysql://{}:{}@{}/{}".format(
    self.db_user,
    self.db_password,
    self.db_host,
    self.db_name
))
session_factory = sessionmaker(self.engine, expire_on_commit=False)
self._session = scoped_session(session_factory)
```

Each worker will use a `scoped_session` that it utilized throughout the backend. For instance, here is a sample for how I generate the token that seems to only be available on _some_ workers.

```
# Check for an existing token for the user
existing_token = self.query(Token).join(User).filter(Token.user_id == user.id).first()

# If the token exists, update some of the fields and return it
if existing_token:
    # If the token is expired, change the token
    if date_parser.parse(existing_token.expires_at) < datetime.utcnow():
        existing_token.token = bcrypt.gensalt().decode('utf-8')

    # Update the expires at
    existing_token.expires_at = (datetime.utcnow() + timedelta(hours=1)).isoformat()
    self.session.commit()
    return existing_token

# If there isn't an existing token, create a new one
token = Token(user_id=user.id, value=bcrypt.gensalt().decode('utf-8'))
self.session.add(token)
self.session.commit()
return token
```

As you can see, I'm making sure that I call `commit` in any case, which should write it to the DB, which it does. When fetching the token from the DB to authenticate the API request, I use this code:

```
# Get the token from the DB
token = DB.session.query(Token).filter(Token.value == token).first()
if not token:
    return None
```

And this code will return None on some workers, but not others.

Does anyone know what I might be doing wrong? If the data is in the DB, shouldn't all of the workers be able to access that data? Any and all help is greatly appreciated! I'm open to adding more information as well, as needed.

**Comment by zzzeek** [comment, marked as answer]
MySQL / InnoDB uses repeatable read isolation by default, and InnoDB also has some quirks of its own.  If those other processes are in transactions that were already begun when the process in question called commit(), they might not see UPDATE changes from that transaction as repeatable read is telling them to show the same row that was just read, even if another transaction emitted an UPDATE.

so you either would reduce isolation level to READ COMMITTED and/or adjust processes so that they themselves only expect to see new changes from the point at which they begun their own transaction.

**Reply by zlshames** [reply]
Thank you! Changing the isolation level fixed the concurrency issue. I appreciate the explanation!
